### PR TITLE
VPN Gateway: Added module

### PR DESCRIPTION
**Pull request by MrMCake**
# Change

- VPN Gateway: Added module
- Added pipelines
- Tested with temp deployed resources (Note: will not deploy with the current dependency pipeline approach due to costs)

Pipeline reference
- [DevOps](https://dev.azure.com/servicescode/infra-as-code-source/_build/results?buildId=206550&view=results) (tested min successfully before)

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update (Wiki)

### CODE REVIEW

**Review by eriqua — COMMENTED** [review]
What's the expensive dependency? Do you have a status badge to attach?

File: .azuredevops/pipelineTemplates/jobs.validateModulePester.yml [hunk 1]
```diff
@@ -172,7 +172,7 @@ jobs:
           testRunTitle: 'Global Module Tests'
           testResultsFormat: NUnit
           testResultsFiles: global-testResults.xml
-          failTaskOnFailedTests: false
+          failTaskOnFailedTests: true
```
Comment (eriqua) [inline .azuredevops/pipelineTemplates/jobs.validateModulePester.yml:175]: Is this from a different PR? what's the change?

File: arm/Microsoft.Network/vpnGateways/.parameters/parameters.json [hunk 2]
```diff
@@ -0,0 +1,65 @@
+{
+    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters": {
+        "name": {
+            "value": "carml-az-vpngw-x-001"
```
Comment (eriqua) [inline arm/Microsoft.Network/vpnGateways/.parameters/parameters.json:6]:
```suggestion
            "value": "<<namePrefix>>-az-vpngw-x-001"
```

File: arm/Microsoft.Network/vpnGateways/.parameters/parameters.json [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        },
+        "virtualHubResourceId": {
+            "value": "/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.Network/virtualHubs/SampleVirtualHub"
```
Comment (eriqua) [inline arm/Microsoft.Network/vpnGateways/.parameters/parameters.json:9]: Can we reference the same virtual Hub created by the virtual hub module? 

```suggestion
            "value": "/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.Network/virtualHubs/<<namePrefix>>-az-vhub-x-001"
```

File: arm/Microsoft.Network/vpnGateways/.parameters/parameters.json [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+                                {
+                                    "id": "/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.Network/virtualHubs/SampleVirtualHub/hubRouteTables/defaultRouteTable"
+                                }
+                            ]
+                        },
+                        "vnetRoutes": {
+                            "staticRoutes": []
+                        }
+                    },
+                    "remoteVpnSiteResourceId": "/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.Network/vpnSites/SampleVpnSite"
```
Comment (eriqua) [inline arm/Microsoft.Network/vpnGateways/.parameters/parameters.json:41]: Do we have a module for this resource?

File: arm/Microsoft.Network/vpnGateways/.parameters/min.parameters.json [hunk 5]
```diff
@@ -0,0 +1,12 @@
+{
+    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters": {
+        "name": {
+            "value": "<<namePrefix>>-az-vpngw-min-001"
+        },
+        "virtualHubResourceId": {
+            "value": "/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.Network/virtualHubs/minHub"
```
Comment (eriqua) [inline arm/Microsoft.Network/vpnGateways/.parameters/min.parameters.json:9]: Can we reference the same virtual Hub created by the virtual hub module? 

```suggestion
            "value": "/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.Network/virtualHubs/<<namePrefix>>-az-vhub-min-001"
```

File: arm/Microsoft.Network/vpnGateways/deploy.bicep [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+@description('Optional. List of all the NAT Rules to associate with the gateway.')
+param natRules array = []
+
+@description('Required. The resource ID of a virtual Hub to connect to. Note: The virtual Hub and Gateway must be deployed into the same location.')
+param virtualHubResourceId string
+
+@description('Optional. BGP settings details.')
+param bgpSettings object = {}
+
+@description('Optional. Enable BGP routes translation for NAT on this VPNGateway.')
```
Comment (eriqua) [inline arm/Microsoft.Network/vpnGateways/deploy.bicep:19]:
```suggestion
@description('Optional. Enable BGP routes translation for NAT on this VPN gateway.')
```

File: arm/Microsoft.Network/vpnGateways/deploy.bicep [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+param enableBgpRouteTranslationForNat bool = false
+
+@description('Optional. Enable Routing Preference property for the Public IP Interface of the VPNGateway.')
```
Comment (eriqua) [inline arm/Microsoft.Network/vpnGateways/deploy.bicep:22]:
```suggestion
@description('Optional. Enable routing preference property for the public IP interface of the VPN gateway.')
```

File: arm/Microsoft.Network/vpnGateways/connections/deploy.bicep [hunk 8]
```diff
@@ -0,0 +1,95 @@
+@description('Required. The name of the VPN connection.')
+param name string
+
+@description('Required. The name of the VPN gateway this VPN connection is associated with.')
+param vpnGatewayName string
+
+@description('Optional. The IPSec Policies to be considered by this connection.')
```
Comment (eriqua) [inline arm/Microsoft.Network/vpnGateways/connections/deploy.bicep:7]:
```suggestion
@description('Optional. The IPSec policies to be considered by this connection.')
```

File: arm/Microsoft.Network/vpnGateways/connections/deploy.bicep [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+param ipsecPolicies array = []
+
+@description('Optional. The Traffic Selector Policies to be considered by this connection.')
```
Comment (eriqua) [inline arm/Microsoft.Network/vpnGateways/connections/deploy.bicep:10]:
```suggestion
@description('Optional. The traffic selector policies to be considered by this connection.')
```

File: arm/Microsoft.Network/vpnGateways/connections/deploy.bicep [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+param trafficSelectorPolicies array = []
+
+@description('Optional. List of all VPN site link connections to the gateway.')
+param vpnLinkConnections array = []
+
+@description('Optional. Routing Configuration indicating the associated and propagated route tables for this connection.')
```
Comment (eriqua) [inline arm/Microsoft.Network/vpnGateways/connections/deploy.bicep:16]:
```suggestion
@description('Optional. Routing configuration indicating the associated and propagated route tables for this connection.')
```

File: arm/Microsoft.Network/vpnGateways/connections/deploy.bicep [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+param routingConfiguration object = {}
+
+@description('Optional. Enable policy-based traffic selectors.')
+param usePolicyBasedTrafficSelectors bool = false
+
+@description('Optional. Use local azure IP to initiate connection.')
```
Comment (eriqua) [inline arm/Microsoft.Network/vpnGateways/connections/deploy.bicep:22]:
```suggestion
@description('Optional. Use local Azure IP to initiate connection.')
```

File: arm/Microsoft.Network/vpnGateways/connections/deploy.bicep [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+param useLocalAzureIpAddress bool = false
+
+@description('Optional. EnableBgp flag.')
```
Comment (eriqua) [inline arm/Microsoft.Network/vpnGateways/connections/deploy.bicep:25]: Could this be better described?

File: arm/Microsoft.Network/vpnGateways/connections/deploy.bicep [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+param enableRateLimiting bool = false
+
+@description('Optional. Enable internet security.')
+param enableInternetSecurity bool = false
+
+@description('Optional. Enable internet security.')
+param enableBgp bool = false
```
Comment (eriqua) [inline arm/Microsoft.Network/vpnGateways/connections/deploy.bicep:32]: Please review description of these 3

File: arm/Microsoft.Network/vpnGateways/natRules/deploy.bicep [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+@description('Optional. An address prefix range of destination IPs on the outside network that source IPs will be mapped to. In other words, your post-NAT address prefix range.')
+param externalMappings array = []
+
+@description('Optional. An address prefix range of source IPs on the inside network that will be mapped to a set of external IPs. In other words, your pre-NAT address prefix range.')
+param internalMappings array = []
+
+@description('Optional. A NAT rule must be configured to a specific VPN Gateway instance. This is applicable to Dynamic NAT only. Static NAT rules are automatically applied to both VPN Gateway instances.')
+param ipConfigurationId string = ''
+
+@description('Optional. The type of NAT rule for VPN NAT. IngressSnat mode (also known as Ingress Source NAT) is applicable to traffic entering the Azure hub\'s Site-to-site VPN gateway. EgressSnat mode (also known as Egress Source NAT) is applicable to traffic leaving the Azure hub\'s Site-to-site VPN gateway.')
```
Comment (eriqua) [inline arm/Microsoft.Network/vpnGateways/natRules/deploy.bicep:16]:
```suggestion
@description('Optional. The type of NAT rule for VPN NAT. IngressSnat mode (also known as Ingress Source NAT) is applicable to traffic entering the Azure hub\'s site-to-site VPN gateway. EgressSnat mode (also known as Egress Source NAT) is applicable to traffic leaving the Azure hub\'s Site-to-site VPN gateway.')
```

File: arm/Microsoft.Network/vpnGateways/natRules/deploy.bicep [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+@allowed([
+  ''
+  'EgressSnat'
+  'IngressSnat'
+])
+param mode string = ''
+
+@description('Optional. The type of NAT rule for VPN NAT. Static one-to-one NAT establishes a one-to-one relationship between an internal address and an external address while Dynamic NAT assigns an IP and port based on availability.')
```
Comment (eriqua) [inline arm/Microsoft.Network/vpnGateways/natRules/deploy.bicep:24]:
```suggestion
@description('Optional. The type of NAT rule for VPN NAT. Static one-to-one NAT establishes a one-to-one relationship between an internal address and an external address while dynamic NAT assigns an IP and port based on availability.')
```

File: arm/Microsoft.Network/vpnGateways/natRules/readme.md [hunk 16]
```diff
@@ -0,0 +1,34 @@
+# VPN Gateways NATRules `[Microsoft.Network/vpnGateways/natRules]`
```
Comment (eriqua) [inline arm/Microsoft.Network/vpnGateways/natRules/readme.md:1]: Please update after deployment file update

File: arm/Microsoft.Network/vpnGateways/readme.md [hunk 17]
```diff
@@ -0,0 +1,93 @@
+# VPN Gateways `[Microsoft.Network/vpnGateways]`
```
Comment (eriqua) [inline arm/Microsoft.Network/vpnGateways/readme.md:1]: Please update readme after deployment file update

File: arm/Microsoft.Network/vpnGateways/connections/readme.md [hunk 18]
```diff
@@ -0,0 +1,67 @@
+# VPN Gateways Connections `[Microsoft.Network/vpnGateways/connections]`
```
Comment (eriqua) [inline arm/Microsoft.Network/vpnGateways/connections/readme.md:1]: Please update readme after deployment file update

File: .azuredevops/pipelineTemplates/jobs.validateModulePester.yml (hunk 1 above)
Comment (MrMCake) [reply .azuredevops/pipelineTemplates/jobs.validateModulePester.yml]: gone it is

File: arm/Microsoft.Network/vpnGateways/.parameters/parameters.json (hunk 4 above)
Comment (MrMCake) [reply arm/Microsoft.Network/vpnGateways/.parameters/parameters.json:41]: No, not yet. It's the next one on the bucket list

Comment (MrMCake) [reply arm/Microsoft.Network/vpnGateways/.parameters/parameters.json:41]: One module at a time ;) Started with vHub, now VPNGateway, next vSite

File: arm/Microsoft.Network/vpnGateways/connections/deploy.bicep (hunk 13 above)
Comment (MrMCake) [reply arm/Microsoft.Network/vpnGateways/connections/deploy.bicep:32]: looks legit

Comment (MrMCake) [reply arm/Microsoft.Network/vpnGateways/connections/deploy.bicep:32]: done

File: arm/Microsoft.Network/vpnGateways/natRules/readme.md (hunk 16 above)
Comment (MrMCake) [reply arm/Microsoft.Network/vpnGateways/natRules/readme.md:1]: done

File: arm/Microsoft.Network/vpnGateways/readme.md (hunk 17 above)
Comment (MrMCake) [reply arm/Microsoft.Network/vpnGateways/readme.md:1]: done

File: arm/Microsoft.Network/vpnGateways/connections/readme.md (hunk 18 above)
Comment (MrMCake) [reply arm/Microsoft.Network/vpnGateways/connections/readme.md:1]: done

File: arm/Microsoft.Network/vpnGateways/.parameters/parameters.json [hunk 19]
```diff
@@ -3,10 +3,10 @@
     "contentVersion": "1.0.0.0",
     "parameters": {
         "name": {
-            "value": "carml-az-vpngw-x-001"
+            "value": "<<namePrefix>>-az-vpngw-x-001"
         },
         "virtualHubResourceId": {
-            "value": "/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.Network/virtualHubs/SampleVirtualHub"
+            "value": "/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.Network/virtualHubs/<<namePrefix>>-az-vhub-x-001"
```
Comment (eriqua) [inline arm/Microsoft.Network/vpnGateways/.parameters/parameters.json:9]: Please update accordingly also the references to "sampleVirtualHub" and "sampleVPNsite" below

Comment (MrMCake) [reply arm/Microsoft.Network/vpnGateways/.parameters/parameters.json:9]: done

**Review by eriqua — APPROVED** [review]
🆗

**Pull request merged**